03_letters/n0npax/main_test.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	"bytes"
+	"strconv"
+	"testing"
+)
+
+func TestLetters(t *testing.T) {
+	testCases := map[string]struct {
+		input    string
+		expected map[rune]int
+	}{
+		"polish": {input: "dzień dobry",
+			expected: map[rune]int{'y': 1, ' ': 1, 'd': 2, 'z': 1, 'i': 1, 'e': 1, 'ń': 1, 'o': 1, 'b': 1, 'r': 1}},
+		"russian": {input: "доброе утро",
+			expected: map[rune]int{'о': 3, 'р': 2, 'т': 1, 'у': 1, ' ': 1, 'е': 1, 'б': 1, 'д': 1}},
+		"english": {input: "good morning",
+			expected: map[rune]int{'o': 3, 'g': 2, 'd': 1, 'm': 1, 'r': 1, 'n': 2, 'i': 1, ' ': 1}},
+		"chineese": {input: "早上好",
+			expected: map[rune]int{'早': 1, '上': 1, '好': 1}},
+	}
+
+	for name, test := range testCases {
+		t.Run(name, helperTestLetter(test.input, test.expected))
+	}
+}
+
+func helperTestLetter(input string, expected map[rune]int) func(*testing.T) {
+	return func(t *testing.T) {
+		result := letters(input)
+		for k, count := range result {
+			if _, ok := expected[k]; !ok {
+				t.Errorf("Found unexpected key: %c", k)
+			}
+			if count != expected[k] {
+				t.Errorf("Incorrect count for key %c. Found: %d, Expected %d", k, count, expected[k])
+			}
+		}
+
+	}
+}
+
+func TestLetterFreqString(t *testing.T) {
+	testCases := map[string]struct {
+		input    letterFreq
+		expected string
+		positive bool
+	}{
+		"latinum":  {input: letterFreq{letter: 'a', count: 2}, expected: "a:2", positive: true},
+		"cyrillic": {input: letterFreq{letter: 'а', count: 2}, expected: "а:2", positive: true},
+		// cyrillic 'а' is different char than latin 'a'
+		"mixed": {input: letterFreq{letter: 'а', count: 2}, expected: "a:2", positive: false},
+	}
+	for name, test := range testCases {
+		t.Run(name, helperTestLetterFreqString(name, test.input, test.expected, test.positive))
```

Comment:
Nit: this is a matter of style, but imo it is better to keep TestFuncs purely self contained module some test fixtures to generate input data. Helpers should not be used to create assertions.

I your case I'd just pull out a bunch of `NotEqual` and `Equal` test cases in two separate test functions and test tables I also recommend using `testify/assert` at this stage.

then your asserts could just by:
```go
for name, tc := range testCases {
     t.Run(name, func(t *testing.T) {
        got :=  tc.input.String()
        assert.Equal(t, tc.expected, got)
    } )
}
```
and in the not equal case:
```go
for name, tc := range testCases {
     t.Run(name, func(t *testing.T) {
        got :=  tc.input.String()
        assert.NotEqual(t, tc.notExpected, got)
    } )
}
```

Reply:
fixed. Thanks for that!